### ISSUE TITLE
Query materialization after GroupBy leads to "unexpected" results

### ISSUE THREAD

**Issue by MaceWindu** [linq2db/linq2db#365]
It is possible to change linq2db behavior to call Enumerable version on GroupBy internally when query materialize and GroupBy call is the last in sequence?
I am requesting it, because in our project we already had several situations when developers write code like that:

```
var dict = db.Table.GroupBy(_ => _.SomeKey).ToDictionary(_ => _.Key, _ => _.ToList());
```

instead of

```
var dict = db.Table.AsEnumerable().GroupBy(_ => _.SomeKey).ToDictionary(_ => _.Key, _ => _.ToList());
```

I think this behavior reflects developer's intension in more proper way.

I'll send pull request with test for current behavior.

**Comment by sdanyliv** [comment]
@ili, it is really fix for current issue?

**Comment by MaceWindu** [comment]
More like a fix for #356

**Comment by ili** [comment]
Yes, you are right

**Comment by jogibear9988** [comment]
@ili Is this really an issue? Should we do smth. special if a Function is the last in sequence?

**Comment by MaceWindu** [comment]
Well, I also don't like it from this PoV, but current behavior also not good. Just yesterday I've got another complain from one for our developers that "GroupBy is broken in linq2db"

**Comment by sdanyliv** [comment]
@MaceWindu, it will be good if you can create PR with unit test that shows problem or new problems.

**Comment by MaceWindu** [comment]
Test is already here https://github.com/linq2db/linq2db/pull/366/

When server-side grouping used, only GROUP BY fields (keys) returned by query and group values list is empty. There is nothing wrong with it if user really wanted to select only keys.
But in that case I think it should be expected from user to use Select to shape result into something more usefull then IGrouping<> values. And if there is no call like that after GroupBy, it means that user wants to work with IGrouping<> values and we need to populate them with values, using Enumerable version of GroupBy.

**Comment by sdanyliv** [comment]
@MaceWindu, it will be ok for now, if we just throw exception in that case?

```'Sequence 'value(LinqToDB.Linq.Table`1[Tests.Model.Person]).GroupBy(keyParam => keyParam.Gender, _ => _)' cannot be converted to SQL'```

Currently there is no easy way in linq2db to process the part of sequence on the client side.

**Comment by MaceWindu** [comment]
Yeah, I think it makes sense. At least we will tell user that he wrote something stupid.

**Comment by sdanyliv** [comment]
Should we always throw exception for that case?
And maybe you have "better message text"?

**Comment by MaceWindu** [comment]
Yep, I have no objections for such option.
About message, maybe something like this:
"You should explicitly specify selected fields for server-side GroupBy() call or add AsEnumerable() call before GroupBy() to perform client-side grouping."
and we can set Exception.HelpLink  to this issue.

**Comment by sdanyliv** [comment]
Fixed in #365

**Comment by filippobottega** [comment]
> Configuration.Linq.GuardGrouping

Thank you! I needed to set Configuration.Linq.GuardGrouping = false. 👍
